File: src/githubclaw/scheduler.py
"""Asyncio-based scheduled event manager.

Persists events to ``~/.githubclaw/scheduled.json`` and checks every 60
seconds for due events.  Supports both recurring and one-shot events.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import logging
import uuid
from dataclasses import asdict, dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable, Coroutine

from githubclaw.constants import (
    MAX_RESCHEDULE_ITERATIONS,
    MAX_SCHEDULED_EVENT_RETRY_COUNT,
    SCHEDULER_CHECK_INTERVAL_SECONDS,
)

logger = logging.getLogger(__name__)

SCHEDULED_JSON = Path.home() / ".githubclaw" / "scheduled.json"
CHECK_INTERVAL_SECONDS = SCHEDULER_CHECK_INTERVAL_SECONDS


@dataclass
class ScheduledEvent:
    """A single scheduled event definition."""

    event_id: str
    repo: str
    trigger_at: str  # ISO-8601 UTC timestamp for next fire
    payload: dict[str, Any]
    one_shot: bool = True
    interval_seconds: int | None = None  # for recurring events
    description: str = ""
    retry_count: int = 0

    def trigger_datetime(self) -> datetime:
        return datetime.fromisoformat(self.trigger_at).replace(tzinfo=UTC)

    def is_due(self, now: datetime | None = None) -> bool:
        now = now or datetime.now(UTC)
        return self.trigger_datetime() <= now


class ScheduledEventManager:
    """Manages scheduled events persisted to disk.

    Call :meth:`start` to begin the asyncio background loop.  Events that
    become due are delivered via the *on_fire* callback provided at
    construction.
    """

    def __init__(
        self,
        on_fire: Callable[[str, dict[str, Any]], Coroutine[Any, Any, None]],
        json_path: str | Path | None = None,
    ) -> None:
        """
        Args:
            on_fire: Async callback ``(repo, payload) -> None`` invoked when
                     a scheduled event fires.
            json_path: Override path for the JSON persistence file (useful
                       for testing).
        """
        self._on_fire = on_fire
        self._json_path = Path(json_path) if json_path else SCHEDULED_JSON
        self._events: list[ScheduledEvent] = []
        self._task: asyncio.Task[None] | None = None
        self._running = False

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------

    def _ensure_dir(self) -> None:
        self._json_path.parent.mkdir(parents=True, exist_ok=True)

    def load(self) -> None:
        """Load scheduled events from disk."""
        self._ensure_dir()
        if not self._json_path.exists():
            self._events = []
            return

        try:
            data = json.loads(self._json_path.read_text(encoding="utf-8"))
            self._events = [ScheduledEvent(**item) for item in data]
            logger.info("Loaded %d scheduled event(s) from %s", len(self._events), self._json_path)
        except (json.JSONDecodeError, TypeError, KeyError) as exc:
            logger.error("Failed to load scheduled events: %s", exc)
            self._events = []

    def save(self) -> None:
        """Persist current events list to disk."""
        self._ensure_dir()
        data = [asdict(ev) for ev in self._events]
        tmp_path = self._json_path.with_suffix(".tmp")
        tmp_path.write_text(json.dumps(data, indent=2), encoding="utf-8")
        tmp_path.replace(self._json_path)

    # ------------------------------------------------------------------
    # CRUD
    # ------------------------------------------------------------------

    def create_event(
        self,
        repo: str,
        trigger_at: datetime,
        payload: dict[str, Any],
        one_shot: bool = True,
        interval_seconds: int | None = None,
        description: str = "",
    ) -> str:
        """Create and persist a new scheduled event.

        Returns:
            The generated ``event_id``.
        """
        # Validate interval_seconds for recurring events
        if not one_shot and interval_seconds is None:
            raise ValueError("Recurring events (one_shot=False) require interval_seconds to be set")
        if interval_seconds is not None and interval_seconds <= 0:
            raise ValueError(f"interval_seconds must be positive, got {interval_seconds}")

        event_id = uuid.uuid4().hex[:12]
        event = ScheduledEvent(
            event_id=event_id,
            repo=repo,
            trigger_at=trigger_at.isoformat(),
            payload=payload,
            one_shot=one_shot,
            interval_seconds=interval_seconds,
            description=description,
        )
        self._events.append(event)
        self.save()
        logger.info("Created scheduled event %s for repo %s (due %s)", event_id, repo, trigger_at)
        return event_id

    def cancel_event(self, event_id: str) -> bool:
        """Cancel a scheduled event by its ID.

        Returns:
            True if the event was found and removed.
        """
        before = len(self._events)
        self._events = [ev for ev in self._events if ev.event_id != event_id]
        removed = len(self._events) < before
        if removed:
            self.save()
            logger.info("Cancelled scheduled event %s", event_id)
        else:
            logger.warning("Scheduled event %s not found for cancellation", event_id)
        return removed

    def list_events(self) -> list[ScheduledEvent]:
        """Return a copy of the current events list."""
        return list(self._events)

    # ------------------------------------------------------------------
    # Firing logic
    # ------------------------------------------------------------------

    async def fire_due_events(self) -> int:
        """Check all events and fire those that are due.

        For one-shot events the entry is removed after firing.
        For recurring events the ``trigger_at`` is advanced by
        ``interval_seconds``.

        Returns:
            The number of events fired.
        """
        now = datetime.now(UTC)
        fired = 0
        remaining: list[ScheduledEvent] = []

        for event in self._events:
            if event.is_due(now):
                try:
                    await self._on_fire(event.repo, event.payload)
                    fired += 1
                    event.retry_count = 0
                    logger.info("Fired scheduled event %s for repo %s", event.event_id, event.repo)
                except Exception:
                    event.retry_count += 1
                    if event.retry_count >= MAX_SCHEDULED_EVENT_RETRY_COUNT:
                        logger.error(
                            "Dead-lettering scheduled event %s for repo %s after %d consecutive failures",
                            event.event_id,
                            event.repo,
                            event.retry_count,
                        )
                        # Do not re-add -- event is dead-lettered
                    else:
                        logger.exception(
                            "Error firing scheduled event %s (retry %d/%d)",
                            event.event_id,
                            event.retry_count,
                            MAX_SCHEDULED_EVENT_RETRY_COUNT,
                        )
                        remaining.append(event)
                    continue

                if event.one_shot:
                    # Do not re-add
                    continue

                # Recurring: advance trigger_at
                if event.interval_seconds:
                    next_trigger = datetime.fromisoformat(event.trigger_at).replace(tzinfo=UTC)
                    # Advance past *now* to avoid rapid repeat fires
                    iterations = 0
                    while next_trigger <= now:
                        next_trigger += timedelta(seconds=event.interval_seconds)
                        iterations += 1
                        if iterations >= MAX_RESCHEDULE_ITERATIONS:
                            logger.warning(
                                "Capped recurring advance at %d iterations for event %s",
                                MAX_RESCHEDULE_ITERATIONS,
                                event.event_id,
                            )
                            break
                    event.trigger_at = next_trigger.isoformat()
                    remaining.append(event)
            else:
                remaining.append(event)

        self._events = remaining
        if fired:
            self.save()
        return fired

    # ------------------------------------------------------------------
    # Background loop
    # ------------------------------------------------------------------

    async def _loop(self) -> None:
        """Background loop that checks for due events every 60 seconds."""
        while self._running:
            try:
                await self.fire_due_events()
            except Exception:
                logger.exception("Unhandled error in scheduled event loop")
            await asyncio.sleep(CHECK_INTERVAL_SECONDS)

    def start(self) -> None:
        """Start the background check loop as an asyncio task."""
        if self._task is not None and not self._task.done():
            logger.warning("Scheduled event loop already running")
            return
        self.load()
        self._running = True
        self._task = asyncio.get_event_loop().create_task(self._loop())
        logger.info("Scheduled event loop started (interval=%ds)", CHECK_INTERVAL_SECONDS)

    async def stop(self) -> None:
        """Stop the background loop gracefully."""
        self._running = False
        if self._task is not None:
            self._task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._task
            self._task = None
        logger.info("Scheduled event loop stopped")

File: src/scheduler.rs
//! Async scheduled event manager.
//!
//! Persists events to a JSON file and checks periodically for due events.
//! Supports both recurring and one-shot events.

use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::constants::*;

fn default_true() -> bool {
    true
}

/// A single scheduled event definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScheduledEvent {
    pub event_id: String,
    pub repo: String,
    /// ISO-8601 UTC timestamp for next fire.
    pub trigger_at: String,
    pub payload: serde_json::Value,
    #[serde(default = "default_true")]
    pub one_shot: bool,
    pub interval_seconds: Option<u64>,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub retry_count: u32,
}

impl ScheduledEvent {
    /// Parse `trigger_at` into a `DateTime<Utc>`.
    pub fn trigger_datetime(&self) -> Option<DateTime<Utc>> {
        DateTime::parse_from_rfc3339(&self.trigger_at)
            .ok()
            .map(|dt| dt.with_timezone(&Utc))
    }

    /// Returns `true` if the event's trigger time is at or before `now`.
    pub fn is_due(&self, now: DateTime<Utc>) -> bool {
        self.trigger_datetime().is_some_and(|dt| dt <= now)
    }
}

/// Manages scheduled events persisted to disk.
///
/// Call [`ScheduledEventManager::load`] to read events from disk.  Events that
/// become due can be discovered via [`ScheduledEventManager::due_events`].
pub struct ScheduledEventManager {
    json_path: PathBuf,
    events: Vec<ScheduledEvent>,
}

impl ScheduledEventManager {
    /// Create a new manager that persists to `json_path`.
    pub fn new(json_path: impl AsRef<Path>) -> Self {
        Self {
            json_path: json_path.as_ref().to_path_buf(),
            events: Vec::new(),
        }
    }

    /// Load scheduled events from disk. If the file does not exist, starts
    /// with an empty list.
    pub fn load(&mut self) -> std::io::Result<()> {
        if let Some(parent) = self.json_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        if !self.json_path.exists() {
            self.events = Vec::new();
            return Ok(());
        }

        let data = std::fs::read_to_string(&self.json_path)?;
        match serde_json::from_str::<Vec<ScheduledEvent>>(&data) {
            Ok(events) => {
                self.events = events;
                Ok(())
            }
            Err(e) => {
                tracing::error!("Failed to parse scheduled events: {}", e);
                self.events = Vec::new();
                Ok(())
            }
        }
    }

    /// Persist current events list to disk using atomic write.
    pub fn save(&self) -> std::io::Result<()> {
        if let Some(parent) = self.json_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let data = serde_json::to_string_pretty(&self.events).map_err(std::io::Error::other)?;

        let tmp_path = self.json_path.with_extension("tmp");
        std::fs::write(&tmp_path, &data)?;
        std::fs::rename(&tmp_path, &self.json_path)?;
        Ok(())
    }

    /// Create and persist a new scheduled event. Returns the generated event ID.
    ///
    /// # Panics
    ///
    /// Panics if `one_shot` is `false` and `interval_seconds` is `None`.
    pub fn create_event(
        &mut self,
        repo: &str,
        trigger_at: DateTime<Utc>,
        payload: serde_json::Value,
        one_shot: bool,
        interval_seconds: Option<u64>,
        description: &str,
    ) -> String {
        if !one_shot && interval_seconds.is_none() {
            panic!("Recurring events (one_shot=false) require interval_seconds to be set");
        }
        if let Some(secs) = interval_seconds {
            assert!(secs > 0, "interval_seconds must be positive, got {secs}");
        }

        let event_id = uuid::Uuid::new_v4().to_string()[..12].to_string();
        let event = ScheduledEvent {
            event_id: event_id.clone(),
            repo: repo.to_string(),
            trigger_at: trigger_at.to_rfc3339(),
            payload,
            one_shot,
            interval_seconds,
            description: description.to_string(),
            retry_count: 0,
        };
        self.events.push(event);
        // Save is best-effort here; callers can check via save() separately
        let _ = self.save();
        event_id
    }

    /// Cancel a scheduled event by its ID. Returns `true` if found and removed.
    pub fn cancel_event(&mut self, event_id: &str) -> bool {
        let before = self.events.len();
        self.events.retain(|ev| ev.event_id != event_id);
        let removed = self.events.len() < before;
        if removed {
            let _ = self.save();
        }
        removed
    }

    /// Return a reference to all current events.
    pub fn list_events(&self) -> &[ScheduledEvent] {
        &self.events
    }

    /// Return references to all events that are due as of `now`.
    pub fn due_events(&self, now: DateTime<Utc>) -> Vec<&ScheduledEvent> {
        self.events.iter().filter(|ev| ev.is_due(now)).collect()
    }

    /// Fire all due events through the provided async callback.
    ///
    /// Returns the number of events successfully fired.  One-shot events are
    /// removed after firing.  Recurring events have their `trigger_at` advanced.
    /// Failed events are retried up to `MAX_SCHEDULED_EVENT_RETRY_COUNT` times
    /// before being dropped (dead-lettered).
    pub async fn fire_due_events_with_callback<F, Fut>(&mut self, callback: F) -> usize
    where
        F: Fn(String, serde_json::Value) -> Fut,
        Fut: std::future::Future<Output = Result<(), String>>,
    {
        let now = chrono::Utc::now();
        let mut fired = 0;
        let mut remaining = Vec::new();

        for event in self.events.drain(..) {
            if event.is_due(now) {
                match callback(event.repo.clone(), event.payload.clone()).await {
                    Ok(()) => {
                        fired += 1;
                        if !event.one_shot {
                            // Advance recurring event
                            if let Some(interval) = event.interval_seconds {
                                let mut e = event;
                                if let Some(mut next) = e.trigger_datetime() {
                                    while next <= now {
                                        next += chrono::Duration::seconds(interval as i64);
                                    }
                                    e.trigger_at = next.to_rfc3339();
                                    e.retry_count = 0;
                                    remaining.push(e);
                                }
                            }
                        }
                        // one_shot: don't re-add
                    }
                    Err(_) => {
                        let mut e = event;
                        e.retry_count += 1;
                        if e.retry_count < MAX_SCHEDULED_EVENT_RETRY_COUNT {
                            remaining.push(e);
                        }
                        // else dead-letter (drop)
                    }
                }
            } else {
                remaining.push(event);
            }
        }

        self.events = remaining;
        if fired > 0 {
            let _ = self.save();
        }
        fired
    }
}

// -----------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Duration;
    use serde_json::json;
    use tempfile::TempDir;

    /// Helper to create a manager backed by a temp directory.
    fn temp_manager() -> (ScheduledEventManager, TempDir) {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("scheduled.json");
        let mgr = ScheduledEventManager::new(&path);
        (mgr, dir)
    }

    // -- ScheduledEvent unit tests --

    #[test]
    fn is_due_with_past_timestamp() {
        let past = Utc::now() - Duration::hours(1);
        let event = ScheduledEvent {
            event_id: "test1".into(),
            repo: "owner/repo".into(),
            trigger_at: past.to_rfc3339(),
            payload: json!({}),
            one_shot: true,
            interval_seconds: None,
            description: String::new(),
            retry_count: 0,
        };
        assert!(event.is_due(Utc::now()));
    }

    #[test]
    fn is_due_with_future_timestamp() {
        let future = Utc::now() + Duration::hours(1);
        let event = ScheduledEvent {
            event_id: "test2".into(),
            repo: "owner/repo".into(),
            trigger_at: future.to_rfc3339(),
            payload: json!({}),
            one_shot: true,
            interval_seconds: None,
            description: String::new(),
            retry_count: 0,
        };
        assert!(!event.is_due(Utc::now()));
    }

    // -- ScheduledEventManager tests --

    #[test]
    fn create_event_adds_to_list_and_saves() {
        let (mut mgr, _dir) = temp_manager();
        let trigger = Utc::now() + Duration::hours(1);
        let id = mgr.create_event(
            "owner/repo",
            trigger,
            json!({"key": "val"}),
            true,
            None,
            "test event",
        );

        assert_eq!(mgr.list_events().len(), 1);
        assert_eq!(mgr.list_events()[0].event_id, id);
        assert_eq!(mgr.list_events()[0].repo, "owner/repo");
        assert_eq!(mgr.list_events()[0].description, "test event");

        // File should exist on disk
        assert!(mgr.json_path.exists());
    }

    #[test]
    fn cancel_event_removes_from_list() {
        let (mut mgr, _dir) = temp_manager();
        let trigger = Utc::now() + Duration::hours(1);
        let id = mgr.create_event("owner/repo", trigger, json!({}), true, None, "");

        assert_eq!(mgr.list_events().len(), 1);
        assert!(mgr.cancel_event(&id));
        assert_eq!(mgr.list_events().len(), 0);
    }

    #[test]
    fn cancel_event_returns_false_for_nonexistent_id() {
        let (mut mgr, _dir) = temp_manager();
        assert!(!mgr.cancel_event("nonexistent_id"));
    }

    #[test]
    fn list_events_returns_all_events() {
        let (mut mgr, _dir) = temp_manager();
        let trigger = Utc::now() + Duration::hours(1);
        mgr.create_event("repo1", trigger, json!({}), true, None, "first");
        mgr.create_event("repo2", trigger, json!({}), true, None, "second");
        mgr.create_event("repo3", trigger, json!({}), true, None, "third");

        assert_eq!(mgr.list_events().len(), 3);
    }

    #[test]
    fn due_events_returns_only_due_events() {
        let (mut mgr, _dir) = temp_manager();
        let now = Utc::now();
        let past = now - Duration::hours(1);
        let future = now + Duration::hours(1);

        mgr.create_event("past-repo", past, json!({}), true, None, "past");
        mgr.create_event("future-repo", future, json!({}), true, None, "future");

        let due = mgr.due_events(now);
        assert_eq!(due.len(), 1);
        assert_eq!(due[0].repo, "past-repo");
    }

    #[test]
    fn load_and_save_roundtrip() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("scheduled.json");

        // Create and save
        {
            let mut mgr = ScheduledEventManager::new(&path);
            let trigger = Utc::now() + Duration::hours(1);
            mgr.create_event(
                "owner/repo",
                trigger,
                json!({"action": "test"}),
                true,
                None,
                "roundtrip",
            );
        }

        // Load in a new manager
        {
            let mut mgr = ScheduledEventManager::new(&path);
            mgr.load().unwrap();
            assert_eq!(mgr.list_events().len(), 1);
            assert_eq!(mgr.list_events()[0].repo, "owner/repo");
            assert_eq!(mgr.list_events()[0].description, "roundtrip");
        }
    }

    #[test]
    fn load_from_nonexistent_file_starts_empty() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("does_not_exist.json");
        let mut mgr = ScheduledEventManager::new(&path);
        mgr.load().unwrap();
        assert!(mgr.list_events().is_empty());
    }

    #[test]
    #[should_panic(expected = "Recurring events (one_shot=false) require interval_seconds")]
    fn recurring_event_requires_interval_seconds() {
        let (mut mgr, _dir) = temp_manager();
        let trigger = Utc::now() + Duration::hours(1);
        mgr.create_event(
            "owner/repo",
            trigger,
            json!({}),
            false,
            None,
            "bad recurring",
        );
    }
}
